test/metabase/test/data.clj

```diff
@@ -212,3 +213,48 @@
   `(-with-temp-db ~database-definition
      (fn [~db-binding]
        ~@body)))
+
+
+(defmacro expect-with-test-db
+  "Identical to `expect` but wraps the entire evaluation in `with-temp-db` so that both expected and actual statements
+   have access to the test data required for the test.
+
+     (expect-with-test-db [_ test-data]
+       (id :venues :id)
+       (sel :one :field [ForeignKey :destination_id] :origin_id (id :checkins :venue_id)))"
+  [test-db-binding expected actual]
+  (let [fn-name (gensym)]
+    `(def ~(vary-meta fn-name assoc :expectation true)
+       (fn []
+         (with-temp-db ~test-db-binding
+           (expectations/doexpect ~expected ~actual))))))
+
+
+;;; ## ---------------------------------------- Test-Data Interface ----------------------------------------
+
+
+(defn with-test-data [i-test-data f]
+  (try
+    (binding [*sel-disable-logging* true]
+      (when (fn? (:before i-test-data))
+        ((:before i-test-data)))
+      (binding [*sel-disable-logging* false]
+        (f)))
+    (finally
+      (binding [*sel-disable-logging* true]
+        (when (fn? (:after i-test-data))
+          ((:after i-test-data)))))))
```

Comment:
Wait a second. Here's an even better way to write it:

```clojure
(defn with-test-data [{:keys [before after]} f]
  (let [no-log (fn [x] (binding [*sel-disable-logging* true] (x)))]
    ((as-> f f
       (if-not (fn? before) f (fn [] (no-log before) (f)))
       (if-not (fn? after)  f (fn [] (try (f)
                                           (finally (no-log after)))))))))
```

This way you can avoid wrapping the whole thing in a try-catch if `after` isn't being used